### PR TITLE
Resume batch file tar download

### LINKED ISSUE #117: Resume download of batch tar file
## Please describe the feature

As a user, I want to be able to resume downloading of multiple files so that I can  save time in case my ongoing download is interrupted.

## Detailed description

This is a follow-up of #116. Details of implementation can be found [here](https://github.com/NBISweden/sda-download-ui/blob/65-options-for-efficient-dataset-download-implementations/docs/tar-download.md).

Note:
Header length is stable but header bytes are not. This means that we cannot support resuming of mid-header interrupted downloads. The single file download implementation already takes this into account and the existing logic can be extended for the bundle case.

- [x] The browser can resume natively the bundle tar file.
- [x] The  untared files are verified that have been resumed correctly.

### PR DESCRIPTION
**Pull request by aaperis**
## Related issue(s) and PR(s)

This PR closes #117  and is a continuation of #116. 

It extends the  batch TAR download endpoint of #116 to support native browser
resume via HTTP Range. 

Again a somewhat lengthy description of the logic:

1. #116 made sure that the TAR's byte layout is deterministic.
2. Here we add a **region map** that is built during the preparation phase. Every byte of the response is tagged as belonging to one of two kinds of region: `static` (in-memory bytes we already have: PAX block, USTAR block, c4gh header, padding, trailer) or `content` (must be fetched from `/files/{id}/content`).
3. A **stable archive (TAR) ETag** is computed from the encoder version, the recipient's PEM checksum, and per-entry `(fileId, content ETag, filePath)`. It deliberately excludes the cached c4gh header bytes so it stays stable for the same recipient and selection. `If-Range` is honored only when this ETag matches.
4. A new `HEAD` handler returns the same headers a `GET` would so that any client that wants to pre-size a transfer can do it cheaply.
5. **Range requests** translate into a walk of the region map starting at the requested offset, slicing static regions and issuing upstream `Range` requests for content regions when only part is needed.

**Note 1**: the encoder version is just a way to document the tar library that was added in #116 since changing this file e.g. modifications in the tar lib that affect block sizes etc (there is no reaosn to do this but who knows) will result in corrupting resumed tars that where originally created by the previous code version.

**Note 2**: Since Crypt4GH headers are size stable but byte unstable, we deny resuming mid-header, similarly as we do with per-file downloads, restarting from scratch instead in such a case. 
Effort was made to minimize such denials by placing a second guard in place. This avoids serving data chunks that end with c4gh header bytes. Instead in such cases it always concatenate with the next chunk and serve them together. This mechanism eliminates the chance that a download is interrupted mid-header if the user presses `pause`, hence the browsers stops gracefully. It may also catch more abrupt interruptions but given that headers are super small as compared to the content, I think this the best that we can do either way.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Testing

- Instructions on how to test

The route's test suite is expanded to cover as much cases as possible (including some edge cases).

To test the resume, follow test instructions of #116 and try to download the 2nd dataset that has a 2Gb file. Pause and resume as much as you like. You can also try to kill the browser and open it again. After the resume picks up and the file is downloaded, validate integrity of files as described in #116.

## Definition of Done checklist

- [x] I have made an effort making the commit history understandable
- [x] I have performed a self-review of my own code and commented any hard-to-understand areas
- [x] Existing tests are passing and I wrote unit tests for new functionality
- [x] My changes generate no new warnings
- [x] All Acceptance Criteria (AC) Fulfilled and checked in the Related issue

### CODE REVIEW

File: frontend/src/app/api/datasets/[datasetId]/download.tar/route.test.ts [hunk 1]
```diff
@@ -616,4 +646,334 @@ describe("GET /api/datasets/[datasetId]/download.tar", () => {
...
+      `bytes ${start}-${end}/${TOTAL}`,
+    );
+
+    const body = await readAll(resp.body);
+    // The 5 bytes the client asked for.
+    expect(body).toEqual(F2_CONTENT.subarray(5, 10));
+    expect(body.byteLength).toBe(5);
+  });
+
+  test("Range ending inside a c4gh header flushes pending at end-of-loop", async () => {
```
Comment (pdmnyberg) [inline frontend/src/app/api/datasets/[datasetId]/download.tar/route.test.ts:900]: I was unable to wrap my head around this test today. But I guess it is fine.

**Review by pdmnyberg — APPROVED** [review]
I wish I had more time to look into this. But I don't see anything that would be too bad for now so I think we can continue. I was not completely able to wrap my head around all the test though.

The functionality seems to work as expected.

File: frontend/src/app/api/datasets/[datasetId]/download.tar/route.test.ts (hunk 1 above)
Comment (aaperis) [reply frontend/src/app/api/datasets/[datasetId]/download.tar/route.test.ts:900]: This test verifies that when a Range request ends inside (or right at the end of) a c4gh header, the byte chunk held in pendingHeader actually gets delivered to the client instead of silently dropped. This avoids promising x bytes and delivering less in such a corner case scenario. So it ensures that the `Content length` is honored. This allows for the specific request to suceed, otherwise the download would fail here due to a content length mismatch.

This ensures that if the request ends with a complete header bytes, this will be delivered and the download is resumable (due to the resume checks regarding mid-header resumes).

If however the byes end mid-header, the download will fail upon resume, but at least it will not fail on first request.
